feat(pageheader): new component [khcp-7898]

kaiarrowood opened the pull request:
# Summary

<!-- Insert a description of the changes in the PR, along with a JIRA ticket reference, if applicable. -->
Add new `AppPageHeader` component for [KHCP-7898](https://konghq.atlassian.net/browse/KHCP-7898).
Also adds support to use new design tokens.

Figma: https://www.figma.com/file/8k70HMMdPj98APLaexYJyR/Header?type=design&node-id=11%3A7494&mode=dev

![image](https://github.com/Kong/public-ui-components/assets/67973710/88a7f84f-192b-4aa0-982e-da5979814764)



## PR Checklist

* [x] **Naming & Structure:** the files and package structure use the conventions outlined in the [Creating a Package docs](https://github.com/Kong/public-ui-components/blob/main/docs/creating-a-package.md).
* [x] **Tests pass:** check the output of all package unit and/or component tests.
  * [ ] If this PR is the result of a bug, test coverage was added accordingly.
* [x] **Functional:** all changes do not break existing APIs, but if so, a BREAKING CHANGE commit is in place to bump the major version.
* [x] **Conventional Commits** all commits follow the conventional commit standards [outlined in the main README](https://github.com/Kong/public-ui-components#committing-changes).
* [x] **Docs:** includes a technically accurate README, and the docs have been updated accordingly based on the changes in this PR.


[KHCP-7898]: https://konghq.atlassian.net/browse/KHCP-7898?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

adamdehaven commented on packages/core/app-layout/src/components/pageHeader/AppPageHeader.vue:
```diff
@@ -0,0 +1,121 @@
+        color: $kui-color-text;
+        font-size: $kui-font-size-70;
+        font-weight: $kui-font-weight-bold;
+        letter-spacing: -0.48px;
```
I'm not sure it makes sense to alter the letter-spacing here; we want a consistent font display throughout all of the apps.

kaiarrowood replied:
Was going off the Figma designs, I'm ok to ignore it if we don't care

adamdehaven replied:
Let's just go with the default for now; gets rid of "custom font" conversations

adamdehaven commented on package.json:
```diff
@@ -24,6 +24,7 @@
     "@commitlint/config-conventional": "^17.6.6",
     "@digitalroute/cz-conventional-changelog-for-jira": "^8.0.1",
     "@evilmartians/lefthook": "^1.4.5",
+    "@kong/design-tokens": "^1.5.1",
```
can you bump this to the latest please?

adamdehaven commented on packages/core/app-layout/docs/page-header.md:
```diff
@@ -0,0 +1,67 @@
+#### `breadcrumbs`
+
+- type: Array as PropType<BreadcrumbItem[]>
+- required: `false`
```
question: why would this prop not be required?

kaiarrowood replied:
You can have a `PageHeader` with no breadcrumb section if you want.

adamdehaven commented on packages/core/app-layout/docs/page-header.md:
```diff
@@ -0,0 +1,67 @@
+
+### Slots
+
+#### `icon`
+
+Content displayed right before the title text, typically an icon.
+
+#### `badge`
+
+Content displayed right after the title text, typically a badge.
```
If these slots can be used for other purposes, then should the slot names be more generic? 

Something like `title-before` and `title-after` ?

adamdehaven commented on packages/core/app-layout/src/components/pageHeader/AppPageHeader.vue:
```diff
@@ -0,0 +1,124 @@
+const props = defineProps({
+  title: {
+    type: String,
+    default: '',
```
nit: if the prop is required, there should not be a default
```suggestion
```

kaiarrowood replied:
I can't keep straight when we explicityly provide a default and when we don't. For example, all the entities components explicitly specify default values even for non-required props, but I'll remove it.

adamdehaven replied:
don't know what people did everywhere else, this is just the way to go

adamdehaven commented on packages/core/app-layout/src/components/pageHeader/AppPageHeader.vue:
```diff
@@ -0,0 +1,124 @@
+  },
+})
+
+const hasBreadcrumbs = computed((): boolean => !!props.breadcrumbs.length)
```
nit: you need to optionally chain here in case the user passes in `undefined`
```suggestion
const hasBreadcrumbs = computed((): boolean => !!props.breadcrumbs?.length)
```

adamdehaven commented on packages/core/app-layout/docs/page-header.md:
```diff
@@ -0,0 +1,67 @@
+## Features
+
+- Reactive updates based on `prop` value changes :rocket:
+- Slottable areas for displaying custom content, icons, etc.
```
suggestion: Can you add an item here about the breadcrumbs (and that they are optional)?

adamdehaven commented on packages/core/app-layout/docs/page-header.md:
```diff
@@ -0,0 +1,67 @@
+## Requirements
+
+- `vue` must be initialized in the host application
+- `@kong/kongponents` must be available as a `dependency` in the host application, along with the package's style imports. [See here for instructions on installing Kongponents](https://kongponents.konghq.com/#globally-install-all-kongponents). Specifically, the following Kongponents must be available:
+  - `KBreadcrumb`
```
I'd remove some text here as KBreadcrumbs also depends on other components
```suggestion
- `@kong/kongponents` must be available as a `dependency` in the host application, along with the package's style imports. [See here for instructions on installing Kongponents](https://kongponents.konghq.com/#globally-install-all-kongponents).
```

kaiarrowood commented:
> A couple things I noticed that seem like they need to be addressed in the UI:
> 
> * The spacing between the title and the badge seems too close when compared to the Figma
>   ![image](https://user-images.githubusercontent.com/2229946/255977287-63cffa1e-74ab-4257-91cb-97542e0c161f.png)
> * This component doesn't appear to be ideally responsive below certain viewport sizes. Can you please add some media queries (using the `kui-breakpoint-*` tokens for media queries) and tweak the layout a bit so that the items properly stack on mobile?
> * If the title text is too long, it send the action menu off of the page. I think we need to add a flex container around the title and action menu (or whatever the right-side content is called) so that the title can wrap to a new line and the action menu remains visible on the right
>   ![image](https://user-images.githubusercontent.com/2229946/255978386-94063dd5-d30c-43f8-bfa3-5c4012b8664a.png)

kaiarrowood commented:
> A couple things I noticed that seem like they need to be addressed in the UI:
> 
> * The spacing between the title and the badge seems too close when compared to the Figma
>   ![image](https://user-images.githubusercontent.com/2229946/255977287-63cffa1e-74ab-4257-91cb-97542e0c161f.png)

Fixed

> * This component doesn't appear to be ideally responsive below certain viewport sizes. Can you please add some media queries (using the `kui-breakpoint-*` tokens for media queries) and tweak the layout a bit so that the items properly stack on mobile?

Added mobile breakpoint styles

> * If the title text is too long, it send the action menu off of the page. I think we need to add a flex container around the title and action menu (or whatever the right-side content is called) so that the title can wrap to a new line and the action menu remains visible on the right
>   ![image](https://user-images.githubusercontent.com/2229946/255978386-94063dd5-d30c-43f8-bfa3-5c4012b8664a.png)

Truncating the title in that scenario instead as agreed by design. Thanks for the help fixing this one @portikM !

@adamdehaven ready for re-review

adamdehaven commented on packages/core/app-layout/src/components/pageHeader/AppPageHeader.vue:
```diff
@@ -0,0 +1,135 @@
+    .page-header-title-wrapper {
+      align-items: baseline;
+      display: flex;
+      min-width: 0;
```
question: why min-width 0?

kaiarrowood replied:
This is an odd one, needed for truncation with flex. See this article @portikM found here: https://css-tricks.com/flexbox-truncated-text/

portikM replied:
Maybe add a comment there explaining why is that `min-width` rule there @kaiarrowood?

adamdehaven commented on packages/core/app-layout/src/components/pageHeader/AppPageHeader.vue:
```diff
@@ -0,0 +1,135 @@
+    margin-bottom: $kui-space-0;
+  }
+
+  @media (max-width: $kui-breakpoint-mobile) {
```
question: why are you using max-width media queries? Styles should be written mobile-first 

Also, if you **must** use max-width, you have to subtract one pixel from the variable or there is an overlap in style rules.

portikM replied:
I'm with Adam on this one - writing styles mobile-first is industry approach to creating responsive UIs. Even though it doesn't currently fit our use-case as mobile screen size support is not on table, it will be one day and then we all will have to use one pattern.

adamdehaven commented on packages/core/app-layout/src/components/pageHeader/AppPageHeader.cy.ts:
```diff
@@ -0,0 +1,84 @@
+
+import AppPageHeader from './AppPageHeader.vue'
+
+describe('<AppPageHeader />', () => {
```
Since the breadcrumb icon slot logic is directly tied to the KBreadcrumbs component, there should be test coverage here to ensure the slot content renders correctly

kaiarrowood replied:
I already added it... https://github.com/Kong/public-ui-components/pull/610/files#diff-37753584988831e9fb98576d5feac57c4849ede0716863f9e9efe4b62e311c98R56
